File: questions/sort/quick_sort.py
# https: // www.geeksforgeeks.org/quick-sort/
# Python program for implementation of Quicksort Sort

# This function takes last element as pivot, places
# the pivot element at its correct position in sorted
# array, and places all smaller (smaller than pivot)
# to left of pivot and all greater elements to right
# of pivot


def partition(arr, low, high):
    i = (low-1)         # index of smaller element
    pivot = arr[high]     # pivot

    for j in range(low, high):
        # If current element is smaller than the pivot
        if arr[j] < pivot:
            # increment index of smaller element
            i = i+1
            arr[i], arr[j] = arr[j], arr[i]

    arr[i+1], arr[high] = arr[high], arr[i+1]
    return (i+1)

# The main function that implements QuickSort
# arr[] --> Array to be sorted,
# low  --> Starting index,
# high  --> Ending index


def quickSort(arr, low, high):
    if low < high:
        # pi is partitioning index, arr[p] is now
        # at right place
        pi = partition(arr, low, high)
        print('pi', pi)

        # Separately sort elements before
        # partition and after partition
        quickSort(arr, low, pi-1)
        quickSort(arr, pi+1, high)


# Function to do Quick sort
# Driver code to test above
arr = [10, 80, 30, 90, 40, 50, 70]
n = len(arr)
quickSort(arr, 0, n-1)
print("Sorted array is:")
for i in range(n):
    print("%d" % arr[i]),

# This code is contributed by Mohit Kumra

File: questions/sort/quick_sort.js
/*
 * Desc: 快速排序
 * File: /quick_sort.js
 * Project: sort
 * File Created: Sunday, 15th March 2020 4:38:32 pm
 * Author: luxuemin2108@gmail.com
 * -----
 * Copyright (c) 2020 Camel Lu
 */

/**
 * @description 找出当前合适的位置，把小于pivot的放在一边，然后替换left的位置
 * @param {*} arr
 * @param {*} low 最低位
 * @param {*} high 最高位
 */
const partition = (arr, low, high) => {
  const pivot = arr[high];
  let left = low - 1;
  for (let i = low; i < high; i++) {
    if (arr[i] < pivot) {
      left += 1;
      if (left !== i) {
        [arr[left], arr[i]] = [arr[i], arr[left]];
      }
    }
  }
  left += 1;
  if (left !== high) {
    [arr[left], arr[high]] = [arr[high], arr[left]];
  }
  return left;
};

let arr = [10, 80, 60, 30, 90, 40, 50, 70];
const high = arr.length;

/**
 * @description 分支
 * @param {*} arr
 * @param {*} low 最低位
 * @param {*} high 最高位
 */
const quickSort = (arr, low, high) => {
  if (low < high) {
    const pivot = partition(arr, low, high);
    quickSort(arr, low, pivot - 1);
    quickSort(arr, pivot + 1, high);
  }
};

quickSort(arr, 0, high - 1);
console.log(arr, arr.length);
